### PR TITLE
[CR] Extract content of Help menu from source code to external file

### PR DESCRIPTION
**Pull request by illi-kun**
I plan to continue this work in further (separate) PRs (if current code and stated target of changes are generally fine).

### CODE REVIEW

File: src/help.cpp [hunk 1]
```diff
@@ -21,8 +21,7 @@ std::vector<std::string> hints;
 std::vector<std::string> text_intro;
 } //namespace
 
-// @todo add required category id as an input parameter
-void load_help()
+void load_help( const std::string required_category )
 {
     std::string path = FILENAMES["help_notes"].c_str();
```
Comment (codemime) [inline src/help.cpp]: `FILENAMES` is a map of `std::string`s. You can assign `path` without constructing the string from `*char`
```c++
const std::string path = FILENAMES["help_notes"]; 
// or even
const auto path = FILENAMES["help_notes"]; 
```

**Comment by BevapDin** [comment]
Stupid question: what's the point of this? What is it needed for? Most of the help text describes the game mechanics, changes to those require changes to the source, which may as well update those text.

I'm not opposed to extracting *data* out of the source files, when it's useful. But I don't see any use here. You might as well extract the name of the game into a separate entry in JSON.

----

Anyhow, you may want to reuse an existing system for loading text from JSON, like the text snippet library instead of implementing yet another one.

File: src/init.cpp [hunk 2]
```diff
@@ -201,6 +201,7 @@ void DynamicDataLoader::initialize()
     type_function_map["INSTRUMENT"] = new StaticFunctionAccessor(&load_ingored_type);
     // loaded earlier.
     type_function_map["colordef"] = new StaticFunctionAccessor(&load_ingored_type);
+    type_function_map["help_notes"] = new StaticFunctionAccessor( &load_ingored_type );
```
Comment (codemime) [inline src/init.cpp]: I couldn't find any object with type `"help_notes"` (neither existing, nor in this PR). Shouldn't it ignore `"help"` instead?

Comment (illi-kun) [inline src/init.cpp]: Yes, you're right.

File: src/help.cpp (hunk 1 above)
Comment (illi-kun) [inline src/help.cpp]: Thanks, I will change this.

**Comment by illi-kun** [comment]
@BevapDin it's not a stupid question at all, that's what I generally want to discuss before I made more changes.

Currently, changing of the content of help menu enforces you to recompile the code but these data are not actually hardly related to the code itself, so I decide that splitting data from code worth the efforts.

It's debatable anyway, so I open for opinions.

> Anyhow, you may want to reuse an existing system for loading text from JSON, like the text snippet library instead of implementing yet another one.

That's what I tried to do at first, but I met the issue: JSON entries are processed/loaded only during the loading of actual game session (i.e. loading of the world). The issue is 'Help' menu is available from the main (starting) screen, and JSON data aren't processed yet.

**Comment by illi-kun** [comment]
__UPD:__ I've added [CR] to the name of PR for gathering more opinions about usefulness of such changes.

**Comment by BevapDin** [comment]
> That's what I tried to do at first, but I met the issue: JSON entries are processed/loaded only during the loading of actual game session (i.e. loading of the world). The issue is 'Help' menu is available from the main (starting) screen, and JSON data aren't processed yet.

Point taken. In that case, there is `load_file` in src/main_menu.cpp, which is used to load credits, title and motd in the main menu. It's a bit different as it expects the files to have properly wrapped lines, but that might actually be useful for the help menu as well.

The help texts aren't structured data, just plain text, storing them in a plain text file should work just fine and you don't have to worry about loading from JSON/ignoring when loading world data etc.

Regarding ignoring help texts in init.cpp: if you store the data (in whatever format you like) in a folder outside of "data/json" (e.g. "data/help") should work fine. The code in init.cpp only looks in "data/json" and "data/mods", so don't have to explicitly ignore that data. And it does not have to follow the common logic used by init.cpp (no "type" member needed).

**Comment by illi-kun** [comment]
> The help texts aren't structured data, just plain text, storing them in a plain text file should work just fine and you don't have to worry about loading from JSON/ignoring when loading world data etc.

I've prefer JSON and not the plain text because of flexibility/expandability of JSON records. For example, "category" helps to sort the arrays of text to sub-categories, and in case of usage of plain text I need to create a dedicated file for each sub-category (```intro.txt```, ```driving.txt```, etc. instead of single ```help.json```) or create some special fields to separate one block of text from another one (i.e. reinvent the JSON format).

Here is example of text which is not movable as-is:
```
text.push_back(string_format(_("\
 In order to assume control of the vehicle, get to a location with \"vehicle controls\" \
 and press %s. Once you are in control, %s accelerates, %s slows or reverses, & %s & %s \
 turn left or right. Diagonals adjust course and speed. You default to cruise control, so \
 the gas/brake adjust the speed which the vehicle will attempt to maintain."),
                                  press_x(ACTION_CONTROL_VEHICLE, "", "").c_str(),
                                  press_x(ACTION_MOVE_N, "", "").c_str(),
                                  press_x(ACTION_MOVE_S, "", "").c_str(),
                                  press_x(ACTION_MOVE_W, "", "").c_str(),
                                  press_x(ACTION_MOVE_E, "", "").c_str()));
```
JSON format allows me to simply add a new array of action names and provide it to code for further processing. In case of working with plain text, I need to replace all '%s' by some special inserts (something like the inlined names of actions embraced by special symbols: ```<ACTION_MOVE_N>```). NB: this is my current vision of situation, it may be changed during the actual implementation of described feature.

> The code in init.cpp only looks in "data/json" and "data/mods", so don't have to explicitly ignore that data. And it does not have to follow the common logic used by init.cpp (no "type" member needed).

It makes sense. Current placeholder in init.cpp is obviously useless and I have to remove it.

**Comment by illi-kun** [comment]
Uh, one more thing concerning this:
> it expects the files to have properly wrapped lines, but that might actually be useful for the help menu as well.

Requirement of preformatted text is pure evil and additional source of headache for translators. Take a look on a simple example -- current state of ```ru.motd```: file contains the explicit statement that length of line more than 72 symbols is not allowed and this requirement was violated in a few lines later:


My conclusion from this example is do not ask a human do something simple but boring, let computer take care of it.

**Comment by BevapDin** [comment]
The action strings are actually the prime example for the point earlier. Adding/removing/changing actions *requires* changing the code. Some additional changes to the contained string literals won't matter at all.

Now you have to invent yet another string replacement system for those texts. For what reason? What could this feature be used for?

> it expects the files to have properly wrapped lines

I mentioned that point *because* it is not optimal. It can be easily changed, we have a word wrapping function for that. However, the help texts are *text*, JSON contains structured data, not plain text. Using JSON to store plain text seems excessive to me.

**Comment by Zireael07** [comment]
Regardless of the format the help files are going to be stored in, this is IMHO a move in a good direction - changing some text in help should not require recompiling.

**Comment by illi-kun** [comment]
> changing some text in help should not require recompiling.

Yeah, that was the reason I've started these changes.

---

> Adding/removing/changing actions requires changing the code.

Yes and no. it's possible to get action names from input data, compare it with actions existing in the game and show key(s) associated with these actions, but it means I need to write it (i.e. make code more complex for getting the same result we already have now).

My current conclusion: this part of the game isn't changed too often and it's not so big, so I probably overestimated the significance of proposed changes initially, and such changes will give minor benefit (if any) and increase the complexity of Help menu's code.

@BevapDin, I really appreciate your participation in our discussion.

I'll close this PR and do some cleanup of help.h/.cpp in a separate PR.

**Pull request closed without merging**